File: site-sample/app/src/main/java/com/huawei/hms/site/demo/DetailSearchActivity.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
 */

package com.huawei.hms.site.demo;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import com.huawei.hms.site.api.SearchResultListener;
import com.huawei.hms.site.api.SearchService;
import com.huawei.hms.site.api.SearchServiceFactory;
import com.huawei.hms.site.api.model.AddressDetail;
import com.huawei.hms.site.api.model.Coordinate;
import com.huawei.hms.site.api.model.CoordinateBounds;
import com.huawei.hms.site.api.model.DetailSearchRequest;
import com.huawei.hms.site.api.model.DetailSearchResponse;
import com.huawei.hms.site.api.model.Poi;
import com.huawei.hms.site.api.model.SearchStatus;
import com.huawei.hms.site.api.model.Site;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;

import androidx.appcompat.app.AppCompatActivity;
import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class DetailSearchActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = "DetailSearchActivity";

    EditText siteIdInput;

    EditText languageInput;

    EditText politicalViewInput;

    TextView resultTextView;

    ImageView sitePhotoView;

    Bitmap bitmap = null;
    public Handler uiHandler = new Handler(new Handler.Callback() {
        @Override
        public boolean handleMessage(Message msg) {
            switch (msg.what) {
                case 0:
                    sitePhotoView.setImageBitmap(bitmap);
                    sitePhotoView.setVisibility(View.VISIBLE);
                    break;
                case 1:
                    sitePhotoView.setImageBitmap(null);
                    sitePhotoView.setVisibility(View.INVISIBLE);
                    break;
                default:
                    break;
            }
            return false;
        }
    });
    private SearchService searchService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail_search);

        searchService = SearchServiceFactory.create(this);

        siteIdInput = findViewById(R.id.edit_text_detail_search_site_id);

        languageInput = findViewById(R.id.edit_text_detail_search_language);

        politicalViewInput = findViewById(R.id.edit_text_detail_search_politicalview);

        findViewById(R.id.button_detail_search).setOnClickListener(this);
        resultTextView = findViewById(R.id.response_detail_search);
        sitePhotoView = findViewById(R.id.search_detail_site_photo_image);
        sitePhotoView.setVisibility(View.INVISIBLE);
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.button_detail_search) {
            execDetailSearch();
        }
    }

    private void execDetailSearch() {
        String siteId = siteIdInput.getText().toString();
        if (TextUtils.isEmpty(siteId) || siteId.getBytes().length > 256) {
            resultTextView.setText("Error : Site Id is empty!");
            return;
        }

        DetailSearchRequest request = new DetailSearchRequest();
        request.setSiteId(siteId);

        String language = languageInput.getText().toString();
        if (!TextUtils.isEmpty(language)) {
            request.setLanguage(language);
        }

        String politicalView = politicalViewInput.getText().toString();
        if (!TextUtils.isEmpty(politicalView)) {
            request.setPoliticalView(politicalView);
        }

        SearchResultListener<DetailSearchResponse> resultListener = new SearchResultListener<DetailSearchResponse>() {
            @Override
            public void onSearchResult(DetailSearchResponse results) {
                Site site;
                if (results == null || (site = results.getSite()) == null) {
                    resultTextView.setText("Result is Empty!");
                    return;
                }
                AddressDetail addressDetail = site.getAddress();
                Coordinate location = site.getLocation();
                Poi poi = site.getPoi();
                CoordinateBounds viewport = site.getViewport();
                StringBuilder response = new StringBuilder("\n");
                response.append("success\n");
                response.append(String.format(
                        "siteId: '%s', Name: %s, FormatAddress: %s, Country: %s, CountryCode: %s, Location: %s, PoiTypes: %s, Viewport: %s \r\n",
                        site.getSiteId(), site.getName(), site.getFormatAddress(),
                        (addressDetail == null ? "" : addressDetail.getCountry()),
                        (addressDetail == null ? "" : addressDetail.getCountryCode()),
                        (location == null ? "" : (location.getLat() + "," + location.getLng())),
                        (poi == null ? "" : Arrays.toString(poi.getPoiTypes())),
                        (viewport == null ? "" : viewport.getNortheast() + "," + viewport.getSouthwest())));

                Switch displaySwitch = findViewById(R.id.detail_search_photo_switch);
                String[] photoUrls = poi != null ? poi.getPhotoUrls() : null;
                String photoUrl = photoUrls != null && photoUrls.length > 0 ? photoUrls[0] : "";
                if (displaySwitch.isChecked() && !TextUtils.isEmpty(photoUrl)) {
                    try {
                        displayPhoto(photoUrl);
                    } catch (Exception e) {
                        e.printStackTrace();
                        resultTextView.setText(response.append("**********************")
                                .append("\r\n")
                                .append("But Get Photo Error : " + e.getMessage())
                                .toString());
                        return;
                    }
                } else {
                    sitePhotoView.setImageBitmap(null);
                }

                resultTextView.setText(response.toString());
                Log.d(TAG, "onDetailSearchResult: " + response.toString());
            }

            @Override
            public void onSearchError(SearchStatus status) {
                resultTextView.setText("Error : " + status.getErrorCode() + " " + status.getErrorMessage());
            }
        };

        searchService.detailSearch(request, resultListener);
    }

    private void displayPhoto(String photoUrl) {
        Request request = new Request.Builder().url(photoUrl).get().build();
        OkHttpClient client = new OkHttpClient();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (response.code() == 200) {
                    InputStream inputStream = response.body().byteStream();
                    bitmap = BitmapFactory.decodeStream(inputStream);
                    uiHandler.sendEmptyMessage(0);
                } else {
                    uiHandler.sendEmptyMessage(1);
                    Log.e(TAG, "get photo from server failed. httpCode=" + response.code());
                }
            }

            @Override
            public void onFailure(Call call, IOException ioe) {
                uiHandler.sendEmptyMessage(1);
                Log.e(TAG, "get photo from server failed.IOException");
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        searchService = null;
    }
}

File: kotlin/app/src/main/java/com/huawei/hms/site/sample/DetailSearchActivity.kt
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
 */
package com.huawei.hms.site.sample

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Bundle
import android.os.Handler
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.Switch
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.gson.Gson
import com.huawei.hms.site.api.SearchResultListener
import com.huawei.hms.site.api.SearchService
import com.huawei.hms.site.api.SearchServiceFactory
import com.huawei.hms.site.api.model.DetailSearchRequest
import com.huawei.hms.site.api.model.DetailSearchResponse
import com.huawei.hms.site.api.model.SearchStatus
import com.huawei.hms.site.api.model.Site
import com.huawei.hms.site.sample.Utils.apiKey
import okhttp3.*
import java.io.IOException
import java.util.*

/**
 * Place Detail Search Example Code
 */
class DetailSearchActivity : AppCompatActivity(), View.OnClickListener {
    companion object {
        private const val TAG = "DetailSearchActivity"
    }

    private lateinit var siteIdInput: EditText
    private lateinit var languageInput: EditText
    private lateinit var resultTextView: TextView
    private lateinit var sitePhotoView: ImageView
    private var bitmap: Bitmap? = null

    // Declare a SearchService object.
    private var searchService: SearchService? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail_search)

        // Instantiate the SearchService object.
        searchService = SearchServiceFactory.create(this, apiKey)
        siteIdInput = findViewById(R.id.edit_text_detail_search_site_id)
        languageInput = findViewById(R.id.edit_text_detail_search_language)
        findViewById<View>(R.id.button_detail_search).setOnClickListener(this)
        resultTextView = findViewById(R.id.response_detail_search)
        sitePhotoView = findViewById(R.id.search_detail_site_photo_image)
        sitePhotoView.visibility = View.INVISIBLE
    }

    override fun onClick(view: View) {
        if (view.id == R.id.button_detail_search) {
            execDetailSearch()
        }
    }

    private fun execDetailSearch() {
        val siteId = siteIdInput.text.toString()
        if (TextUtils.isEmpty(siteId) || siteId.toByteArray().size > 256) {
            resultTextView.text = "Error : Site Id is empty!"
            return
        }

        // Create a request body.
        val request = DetailSearchRequest()
        request.setSiteId(siteId)
        val language = languageInput.text.toString()
        if (!TextUtils.isEmpty(language)) {
            request.setLanguage(language)
        }
        request.isChildren = findViewById<Switch>(R.id.switch_detail_search_children).isChecked

        // Create a search result listener.
        val resultListener: SearchResultListener<DetailSearchResponse> = object : SearchResultListener<DetailSearchResponse> {
            // Return search results upon a successful search.
            override fun onSearchResult(results: DetailSearchResponse?) {
                var site: Site? = null
                if (results == null || results.getSite().also { site = it } == null) {
                    resultTextView.text = "Result is Empty!"
                    return
                }
                val addressDetail = site?.getAddress()
                val location = site?.getLocation()
                val poi = site?.getPoi()
                val viewport = site?.getViewport()
                val response = StringBuilder("\n")
                response.append("success\n")
                response.append(String.format(
                        "siteId: '%s', Name: %s, FormatAddress: %s, Country: %s, CountryCode: %s, Location: %s, PoiTypes: %s, Viewport: %s, ",
                        site?.getSiteId(), site?.getName(), site?.getFormatAddress(),
                        if (addressDetail == null) "" else addressDetail.getCountry(),
                        if (addressDetail == null) "" else addressDetail.getCountryCode(),
                        if (location == null) "" else location.getLat().toString() + "," + location.getLng(),
                        if (poi == null) "" else Arrays.toString(poi.getPoiTypes()),
                        if (viewport == null) "" else "northeast{lat=" + viewport.getNortheast().getLat().toString() + ", lng=" + viewport.getNortheast().getLng().toString() + "},"
                                + "southwest{lat=" + viewport.getSouthwest().getLat().toString() + ", lng=" + viewport.getSouthwest().getLng().toString() + "}"))

                if (poi != null) {
                    val g = Gson()
                    val jsonString = g.toJson(poi.getChildrenNodes())
                    response.append(String.format("children: %s\n\n", jsonString))
                }
                val displaySwitch = findViewById<Switch>(R.id.detail_search_photo_switch)
                val photoUrls = poi?.getPhotoUrls()
                val photoUrl = if (photoUrls != null && photoUrls.isNotEmpty()) photoUrls[0] else ""
                if (displaySwitch.isChecked && !TextUtils.isEmpty(photoUrl)) {
                    try {
                        displayPhoto(photoUrl)
                    } catch (e: Exception) {
                        e.printStackTrace()
                        resultTextView.text = response.append("**********************")
                                .append("\r\n")
                                .append("But Get Photo Error : " + e.message)
                                .toString()
                        return
                    }
                } else {
                    sitePhotoView.setImageBitmap(null)
                }
                resultTextView.text = response.toString()
                Log.d(TAG, "onDetailSearchResult: $response")
            }

            // Return the result code and description upon a search exception.
            override fun onSearchError(status: SearchStatus) {
                resultTextView.text = "Error : ${status.getErrorCode()}  ${status.getErrorMessage()}"
            }
        }
        // Call the place details search API.
        searchService?.detailSearch(request, resultListener)
    }

    val uiHandler = Handler(Handler.Callback { msg ->
        when (msg.what) {
            0 -> {
                sitePhotoView.setImageBitmap(bitmap)
                sitePhotoView.visibility = View.VISIBLE
            }
            1 -> {
                sitePhotoView.setImageBitmap(null)
                sitePhotoView.visibility = View.INVISIBLE
            }
            else -> {
            }
        }
        false
    })

    private fun displayPhoto(photoUrl: String) {
        val request = Request.Builder().url(photoUrl).get().build()
        val client = OkHttpClient()
        client.newCall(request).enqueue(object : Callback {
            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                if (response.code() == 200) {
                    val inputStream = response.body()?.byteStream()
                    bitmap = BitmapFactory.decodeStream(inputStream)
                    uiHandler.sendEmptyMessage(0)
                } else {
                    uiHandler.sendEmptyMessage(1)
                    Log.e(TAG, "get photo from server failed. httpCode=${response.code()}")
                }
            }

            override fun onFailure(call: Call, ioe: IOException) {
                uiHandler.sendEmptyMessage(1)
                Log.e(TAG, "get photo from server failed.IOException")
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        searchService = null
    }
}